### PR TITLE
IGNITE-21908 Add metrics of distribution among stripes in disruptor

### PR DESCRIPTION
**Pull request by vldpyatkov**
https://issues.apache.org/jira/browse/IGNITE-21908

### CODE REVIEW

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java [hunk 1]
```diff
@@ -59,18 +60,34 @@ public class DistributionMetric extends AbstractMetric implements CompositeMetri
     /** Bounds of measurements. */
     private final long[] bounds;
 
+    /** Value string formatter or {@code null} to use default formatter. */
```
Comment (sk0x50) [inline modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java:63]: It seems to me, that this change should be applied to all metric classes: `IntMetric`, `IntGaguge` etc

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java [hunk 2]
```diff
@@ -59,18 +60,34 @@ public class DistributionMetric extends AbstractMetric implements CompositeMetri
...
+    private final Function<long[], String> stringFormatter;
+
     /** List of scalar metrics. */
     private volatile List<Metric> scalarMetrics = null;
 
     /**
-     * The constructor.
+     * Constructor.
```
Comment (sk0x50) [inline modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java:70]: You removed the article but added it a few lines below :)

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MovingAverageMetric.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.function.DoubleFunction;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * The metric calculates the average value for the last several operations.
+ */
+public class MovingAverageMetric extends AbstractMetric implements DoubleMetric {
+    /** Default rate. */
```
Comment (sk0x50) [inline modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MovingAverageMetric.java:28]: I think this is not a rate. `window size`?

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MovingAverageMetric.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+package org.apache.ignite.internal.metrics;
+
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.function.DoubleFunction;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * The metric calculates the average value for the last several operations.
+ */
+public class MovingAverageMetric extends AbstractMetric implements DoubleMetric {
```
Comment (sk0x50) [inline modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MovingAverageMetric.java:27]: Let's rename this class as `SimpleMovingAverage`

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java [hunk 5]
```diff
@@ -1247,17 +1251,20 @@ private void initPools(final NodeOptions opts) {
...
-                "JRaft-FSMCaller-Disruptor-" + ownFsmCallerExecutorDisruptorConfig.getThreadPostfix(),
+                "JRaft-FSMCaller-Disruptor" + ownFsmCallerExecutorDisruptorConfig.getThreadPostfix(),
                 opts.getRaftOptions().getDisruptorBufferSize(),
                 () -> new FSMCallerImpl.ApplyTask(),
                 ownFsmCallerExecutorDisruptorConfig.getStripes(),
                 false,
-                false
+                false,
+                null
+                //opts.getRaftMetrics().disruptorMetrics("JRaft-FSMCaller-Disruptor")
```
Comment (sk0x50) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java:1267]: should it be removed?

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MovingAverageMetric.java (hunk 4 above)
Comment (vldpyatkov) [reply modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MovingAverageMetric.java:27]: Renamed

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java (hunk 2 above)
Comment (vldpyatkov) [reply modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java:70]: This one is just a call to another constructor.
I am not sure about the usage of article hire, but I think the only constructor can be marked with the artical 'the'.

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java (hunk 1 above)
Comment (vldpyatkov) [reply modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java:63]: Maybe we will do it when we start to use them.
Currently, it just adds code to review here.

File: modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageManagerImplTest.java [hunk 6]
```diff
@@ -110,7 +111,8 @@ void setUp(
 
         var raftGroupEventsClientListener = new RaftGroupEventsClientListener();
 
-        raftManager = new Loza(clusterService, raftConfiguration, workDir.resolve("loza"), clock, raftGroupEventsClientListener);
+        raftManager = new Loza(clusterService, new MetricManager(), raftConfiguration, workDir.resolve("loza"), clock,
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageManagerImplTest.java:114]: Added a 'noOp' implementation to use in tests.

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java (hunk 1 above)
Comment (sk0x50) [reply modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java:63]: Honestly, I don't see any difference between `IngGaguge` and `LongGaguge`, and therefore this improvement should be added to both, I think. Please create a ticket to address this.

Comment (vldpyatkov) [reply modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java]: I removed all the string formatters in metriics.
We will add them in the future after discussion.

**Review by sk0x50 — APPROVED** [review]

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/SimpleMovingAverage.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    ConcurrentLinkedDeque<Double> queue = new ConcurrentLinkedDeque<>();
+
+    /**
+     * The constructor.
+     *
+     * @param name Name.
+     * @param desc Description.
+     * @param stringFormatter String formatter to get a readable value.
+     */
+    public SimpleMovingAverage(String name, @Nullable String desc, @Nullable DoubleFunction<String> stringFormatter) {
```
Comment (kgusakov) [inline modules/metrics/src/main/java/org/apache/ignite/internal/metrics/SimpleMovingAverage.java:47]: As I understand we agreed to remove formatter for now?

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/metrics/src/main/java/org/apache/ignite/internal/metrics/SimpleMovingAverage.java (hunk 7 above)
Comment (vldpyatkov) [reply modules/metrics/src/main/java/org/apache/ignite/internal/metrics/SimpleMovingAverage.java:44]: Fixed.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/impl/JraftServerImpl.java [hunk 8]
```diff
@@ -328,7 +333,8 @@ public CompletableFuture<Void> startAsync() {
                     ApplyTask::new,
                     opts.getStripes(),
                     false,
-                    false
+                    false,
+                    opts.getRaftMetrics().disruptorMetrics("JRaft-FSMCaller-Disruptor")
```
Comment (kgusakov) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/server/impl/JraftServerImpl.java:337]: Not sure, that I understand the idea - why we set here "JRaft-FSMCaller-Disruptor" name and then convert it to "jraft.fsmcaller.disruptor" immediately? Maybe use the "jraft.fsmcaller.disruptor" here directly?

**Review by kgusakov — CHANGES_REQUESTED** [review]

**Review by kgusakov — APPROVED** [review]

**Pull request merged**